### PR TITLE
Mark `_LayoutBuilderElement` as always clean

### LINKED ISSUE #154060: App crashes with `"BuildOwner.scheduleBuildFor() called inappropriately." in callback for "ext.flutter.reassemble"` when hot-reloading multiple times
### Steps to reproduce

1. Create an app with the provided code sample.
2. Launch the app on iOS simulator.
3. Trigger a Hot Reload. (No error should trigger at this stage.)
4. Trigger another Hot Reload.
5. Observe the below error.

It seems the issue is with using a global (or provider) `GoRouter` while having `extendBodyBehindAppBar` set to `true` on the `HomeView`'s `Scaffold` widget.

### Expected results

No error should be thrown when using `GoRouter` and `extendBodyBehindAppBar`.

### Actual results

After the first Hot Reload, every consecutive Hot Reload triggers the following error:
```
The following assertion was thrown during a service extension callback for "ext.flutter.reassemble":
BuildOwner.scheduleBuildFor() called inappropriately.
The BuildOwner.scheduleBuildFor() method should only be called while the buildScope() method is actively rebuilding the widget tree.
```

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const AppScreen());
}

// AppScreen
class AppScreen extends StatelessWidget {
  const AppScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: router,
    );
  }
}

// HomeView
class HomeView extends StatelessWidget {
  const HomeView({super.key});

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Scaffold(
          body: SizedBox(
            width: double.maxFinite,
            child: Container(color: Colors.pink),
          ),
          extendBodyBehindAppBar: true,
        ),
      ],
    );
  }
}

// SignInView
class SignInView extends StatelessWidget {
  const SignInView({super.key});

  @override
  Widget build(BuildContext context) {
    return Container(color: Colors.green);
  }
}

// Router
final router = GoRouter(
  initialLocation: '/',
  errorBuilder: (_, state) => Container(color: Colors.red),
  redirect: (_, state) {
    print(state.matchedLocation);
    return (state.matchedLocation != '/sign-in') ? '/sign-in' : null;
  },
  routes: [
    GoRoute(
      name: 'home',
      path: '/',
      builder: (_, state) => HomeView(
        key: state.pageKey,
      ),
      routes: [
        GoRoute(
          name: 'signIn',
          path: 'sign-in',
          builder: (_, state) => SignInView(
            key: state.pageKey,
          ),
        ),
      ],
    ),
  ],
);
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

Console log: [Console log.txt](https://github.com/user-attachments/files/16738224/Console.log.txt)

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.0, on macOS 14.6.1 23G93 darwin-x64, locale en-US)
    • Flutter version 3.24.0 on channel stable at /.../flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 80c2e84975 (3 weeks ago), 2024-07-30 23:06:49 +0700
    • Engine revision b8800d88be
    • Dart version 3.5.0
    • DevTools version 2.37.2

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /.../Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Network resources
    • All expected network resources are available.
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes https://github.com/flutter/flutter/issues/154060

The error message doesn't make sense to me since one can call `setState` during the idle phase, and I'm not sure what this is guarding against without the right error message.

In the case of #154060 the layout builder was never laid out:
```
├─child 1: _RenderLayoutBuilder#7c319 NEEDS-LAYOUT NEEDS-PAINT
│   creator: LayoutBuilder ← _BodyBuilder ← MediaQuery ←
│     LayoutId-[<_ScaffoldSlot.body>] ← CustomMultiChildLayout ←
│     _ActionsScope ← Actions ← AnimatedBuilder ← DefaultTextStyle ←
│     AnimatedDefaultTextStyle ← _InkFeatures-[GlobalKey#1f6eb ink
│     renderer] ← NotificationListener<LayoutChangedNotification> ← ⋯
│   parentData: offset=Offset(0.0, 0.0); id=_ScaffoldSlot.body
│   constraints: MISSING
│   size: MISSING
```
So https://github.com/flutter/flutter/pull/154681 doesn't really fix #154060 since the layout callback cannot be run without a set of valid constraints. 

Before the `BuildScope` change all `_inDirtyList` flags were unset after the `BuildOwner` finishes rebuilding the widget tree, so `LayoutBuilder._inDirtyLst` is always set to false after a frame even for layout builders that were never laid out.
With the `BuildScope` change, `LayoutBuilder` has its own `BuildScope` which is only flushed after LayoutBuilder gets its constraints. 

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by goderbauer** [comment]
After some digging around, I think this assert is supposed to guard against calling `scheduleBuildFor` multiple times for the same element. `Element.markNeedsBuild` (which is called by `setState`) guards its call to `scheduleBuildFor` on whether the element is dirty. If it isn't dirty, it returns right there and we skip all the build scheduling steps:

https://github.com/flutter/flutter/blob/84757af9f809f6709d781779452f8b72063f9640/packages/flutter/lib/src/widgets/framework.dart#L5210-L5214

Can you help me understand why it should conceptually be acceptable to schedule multiple builds for the same element?

(I do agree that the ErrorHint is bogus.)

**Comment by LongCatIsLooong** [comment]
> After some digging around, I think this assert is supposed to guard against calling `scheduleBuildFor` multiple times for the same element. `Element.markNeedsBuild` (which is called by `setState`) guards its call to `scheduleBuildFor` on whether the element is dirty. If it isn't dirty, it returns right there and we skip all the build scheduling steps:
> 
> https://github.com/flutter/flutter/blob/84757af9f809f6709d781779452f8b72063f9640/packages/flutter/lib/src/widgets/framework.dart#L5210-L5214
> 
> Can you help me understand why it should conceptually be acceptable to schedule multiple builds for the same element?
> 
> (I do agree that the ErrorHint is bogus.)

Oh that makes sense thanks for looking into it. 

I think it's the `LayoutBuilderElement` that's triggering the error but it doesn't actually schedule multiple builds. If the render object couldn't do layout then the element stays dirty and `_inDirtyList` is always true. I'll move the check to the `BuildScope` class then.

**Comment by LongCatIsLooong** [comment]
Alternative 1:
```diff
-  void buildScope(Element context, [ VoidCallback? callback ]) {
-    final BuildScope buildScope = context.buildScope;
+  void buildScope(Element context, [ VoidCallback? callback, BuildScope? buildScopeOverride ]) {
+    final BuildScope buildScope = buildScopeOverride ?? context.buildScope;
     if (callback == null && buildScope._dirtyElements.isEmpty) {
       return;
     }
modified   packages/flutter/lib/src/widgets/layout_builder.dart
@@ -78,6 +78,31 @@ abstract class ConstrainedLayoutBuilder<ConstraintType extends Constraints> exte
   // updateRenderObject is redundant with the logic in the LayoutBuilderElement below.
 }
 
+class _BuildScopeOwner extends ProxyWidget {
+  const _BuildScopeOwner(this.buildScope, Widget child) : super(child: child);
+
+  final BuildScope buildScope;
+
+  @override
+  Element createElement() => _ElementWithBuildScope(this);
+}
+
+class _ElementWithBuildScope extends ComponentElement {
+  _ElementWithBuildScope(_BuildScopeOwner super.widget);
+
+  @override
+  BuildScope get buildScope => (widget as _BuildScopeOwner).buildScope;
+
+  @override
+  void update(Widget newWidget) {
+    super.update(newWidget);
+    rebuild(force: true);
+  }
+
+  @override
+  Widget build() => (widget as _BuildScopeOwner).child;
+}
+
 class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderObjectElement {
   _LayoutBuilderElement(ConstrainedLayoutBuilder<ConstraintType> super.widget);
 
@@ -86,9 +111,6 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
 
   Element? _child;
 
-  @override
-  BuildScope get buildScope => _buildScope;
-
   late final BuildScope _buildScope = BuildScope(scheduleRebuild: _scheduleRebuild);
 
   // To schedule a rebuild, markNeedsLayout needs to be called on this Element's
@@ -100,7 +122,6 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
     if (_deferredCallbackScheduled) {
       return;
     }
-
     final bool deferMarkNeedsLayout = switch (SchedulerBinding.instance.schedulerPhase) {
       SchedulerPhase.idle || SchedulerPhase.postFrameCallbacks => true,
       SchedulerPhase.transientCallbacks || SchedulerPhase.midFrameMicrotasks || SchedulerPhase.persistentCallbacks => false,
@@ -208,6 +229,7 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
           ),
         );
       }
+      built = _BuildScopeOwner(_buildScope, built);
       try {
         _child = updateChild(_child, built, null);
         assert(_child != null);
@@ -233,7 +255,7 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
     final VoidCallback? callback = _needsBuild || (constraints != _previousConstraints)
       ? updateChildCallback
       : null;
-    owner!.buildScope(this, callback);
+    owner!.buildScope(this, callback, _buildScope);
   }
 ```

File: packages/flutter/lib/src/widgets/layout_builder.dart [hunk 1]
```diff
@@ -158,7 +158,10 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
 
   @override
   void markNeedsBuild() {
-    super.markNeedsBuild();
+    // Calling super.markNeedsBuild is not needed. This Element does not need
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/layout_builder.dart:161]: If this is the case, will this class's performRebuild be called at all? if the assumption is this element is never dirty, we should probably assert the performRebuild not getting called at all

**Review by chunhtai — COMMENTED** [review]
I am still trying to wrap my head around this. Can you explain a bit the order of things that happens that cause the assertion to be thrown?

File: packages/flutter/lib/src/widgets/layout_builder.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/layout_builder.dart:161]: Yes, when the parent Element calls `updateChild` (i.e. in treewalk).

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/layout_builder.dart:161]: ah I see, I just found out the `RenderObjectElement` overrides the `update`

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -2911,15 +2911,14 @@ class BuildOwner {
                   'The dirty list for the current build scope is: ${buildScope._dirtyElements}',
         );
       }
-      // When reactivating an inactivate Element, _scheduleBuildFor should only be
-      // called within _flushDirtyElements.
       if (!_debugBuilding && element._inDirtyList) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/framework.dart:2914]: This check assume repeated call happens within the same frame, which is no longer the case in the layoutbuilder.

WDYT about updating the check to take into account that it is not guarantee the build will be flush every frame.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/framework.dart:2914]: One possible way is to somehow only enforce this on root build scope

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/framework.dart:2914]: actually know I think of it the current pr makes more sense. there is just nothing to do for rebuild for layoutBuilder

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/framework.dart:2914]: Ideally the assert should still hold: https://github.com/flutter/flutter/pull/154681

**Comment by LongCatIsLooong** [comment]
update `Element.dirty` docs

**Pull request merged**